### PR TITLE
Add Jib integration

### PR DESCRIPTION
**Pull request by eddumelendez**
[Jib](https://github.com/GoogleContainerTools/jib) is a java library to build container images.

Provide `JibDockerClient` and `JibImage` to be used along with `new GenericContainer()`. Example:

```java
GenericContainer<?> busybox = new GenericContainer<>(
    new JibImage(
        "busybox:1.35",
        jibContainerBuilder -> jibContainerBuilder.setEntrypoint("echo", "Hello World")
    )
)
```

### CODE REVIEW

File: modules/jib/src/test/java/org/testcontainers/jib/JibTest.java
Comment (bsideup) [inline modules/jib/src/test/java/org/testcontainers/jib/JibTest.java:37]: This could be `new GenericContainer`

File: modules/jib/src/main/java/org/testcontainers/jib/TcImageDetails.java
Comment (bsideup) [inline modules/jib/src/main/java/org/testcontainers/jib/TcImageDetails.java:10]: does this need to be public?

File: modules/jib/build.gradle
Comment (bsideup) [inline modules/jib/build.gradle:1]: WDYT about it to `core`?

File: modules/jib/src/main/java/org/testcontainers/jib/TcDockerClient.java
Comment (bsideup) [inline modules/jib/src/main/java/org/testcontainers/jib/TcDockerClient.java:43]: We should probably pipe here, to avoid OOMs on large images

File: modules/jib/src/main/java/org/testcontainers/jib/TcDockerClient.java
Comment (bsideup) [inline modules/jib/src/main/java/org/testcontainers/jib/TcDockerClient.java:29]: `JibDockerClient` maybe?

File: modules/jib/src/test/java/org/testcontainers/jib/JibTest.java
Comment (bsideup) [inline modules/jib/src/test/java/org/testcontainers/jib/JibTest.java:35]: Shouldn't we be using `getTargetImage` here?

File: modules/jib/src/main/java/org/testcontainers/jib/JibDockerClient.java
Comment (eddumelendez) [inline modules/jib/src/main/java/org/testcontainers/jib/JibDockerClient.java:69]: currently, pulling has been added here but in order to let jib pull docker client should not be provided 

https://github.com/GoogleContainerTools/jib/blob/9027c8dafb8ee12b9d37e1521caf62dfbcc3d374/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java#L234-L249

Comment (eddumelendez) [inline modules/jib/src/main/java/org/testcontainers/jib/JibDockerClient.java:69]: currently, pulling has been added here but in order to let jib pull docker client should not be provided 

https://github.com/GoogleContainerTools/jib/blob/9027c8dafb8ee12b9d37e1521caf62dfbcc3d374/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java#L234-L249

Comment (eddumelendez) [inline modules/jib/src/main/java/org/testcontainers/jib/JibDockerClient.java:69]: currently, pulling has been added here but in order to let jib pull docker client should not be provided 

https://github.com/GoogleContainerTools/jib/blob/9027c8dafb8ee12b9d37e1521caf62dfbcc3d374/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java#L234-L249

/cc @bsideup

Comment (bsideup) [reply modules/jib/src/main/java/org/testcontainers/jib/JibDockerClient.java:69]: idea: we could make this class a singleton and then use TC's abstraction. For example, you can pull with `RemoteDockerImage` (will support `ImageNameSubstitutor`)

File: modules/jib/build.gradle
Comment (kiview) [reply modules/jib/build.gradle:1]: +1 from me for having this in `core`.

**Comment by eddumelendez** [comment]
if `Jib.from(String)` is used the image is downloaded using Jib's http client. `Jib.from(DockerDaemonImage)` uses new `JibDockerClient#inspect` which will use `RemoteDockerImage(DockerImageName)` to allow image name substitution.

**Comment by eddumelendez** [comment]
Also, classes now are in `core` module and the upgrade of jackson library is required in order to make it work with Jib.

File: core/src/main/resources/META-INF/services/com.google.cloud.tools.jib.api.DockerClient
Comment (bsideup) [inline core/src/main/resources/META-INF/services/com.google.cloud.tools.jib.api.DockerClient:1]: I don't think we should implement the SPI. It is too implicit, **especially** given that `supported` returns `true` unconditionally.

I think we should just provide a singleton instance anyone can use with Jib explicitly.

File: core/src/test/java/org/testcontainers/containers/JibTest.java
Comment (bsideup) [inline core/src/test/java/org/testcontainers/containers/JibTest.java:24]: nit: this should probably just be `throws Exception` :D

File: core/src/test/java/org/testcontainers/containers/JibTest.java
Comment (bsideup) [inline core/src/test/java/org/testcontainers/containers/JibTest.java:33]: I think we would really benefit from something like `new GenericContainer<>(new JibImage(jib -> jib.from(image).setEntrypoint(...)))`, where `class JibImage implements Future<String>`. Otherwise, it doesn't feel actually integrated. WDYT?

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/JibDockerClient.java:46]: Have you considered streaming operations here? `ImageTarball` could be pretty big and cause OOMs

File: core/src/main/resources/META-INF/services/com.google.cloud.tools.jib.api.DockerClient
Comment (eddumelendez) [reply core/src/main/resources/META-INF/services/com.google.cloud.tools.jib.api.DockerClient]: The reason to use the SPI was to delegate everything under the same behavior. By default, the Jib implementation will use docker cli. I was looking for a way to hide it.

File: core/src/test/java/org/testcontainers/containers/JibTest.java
Comment (eddumelendez) [reply core/src/test/java/org/testcontainers/containers/JibTest.java]: It would be great if we can set the docker client to the JibContainer and use tc abstractions there. I agree with this approach indeed I think I overthought this approach.

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/containers/JibDockerClient.java:56]: thanks for pointing out this, still looking at it.

File: core/build.gradle
Comment (bsideup) [inline core/build.gradle:71]: btw this should be provided, we don't want to have Jib as a mandatory dependency

File: core/build.gradle
Comment (bsideup) [inline core/build.gradle:58]: we forced it for reason, see https://github.com/testcontainers/testcontainers-java/pull/3602

Why does Jib require a bump?

Comment (eddumelendez) [reply core/build.gradle:58]: it fails because of this without bumping the jackson version

```
Caused by: java.lang.ClassNotFoundException: com.fasterxml.jackson.databind.ser.std.ToStringSerializerBase
	at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 23 more
```

Comment (eddumelendez) [reply core/build.gradle:58]: I have forced `com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.8.8`

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/containers/JibDockerClient.java]: it has been implemented but depends on https://github.com/docker-java/docker-java/pull/1982

File: build.gradle
Comment (kiview) [inline build.gradle:51]:
```suggestion
```
Not necessary anymore.

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/JibDockerClient.java:65]: I think https://github.com/testcontainers/testcontainers-java/pull/5623#discussion_r934976478 also holds true here and would similarly require a `saveImageAsyncCmd` if we want to implement it cleanly.

File: core/src/test/java/org/testcontainers/containers/JibTest.java
Comment (kiview) [inline core/src/test/java/org/testcontainers/containers/JibTest.java:33]: Needs to be added to the docs as well, so users understand how to use it.

File: docs/features/jib.md
Comment (kiview) [inline docs/features/jib.md:11]:
```suggestion
The `JibDockerClient` should be used as follows
```

File: docs/features/jib.md
Comment (kiview) [inline docs/features/jib.md:15]: Let's add a documentation paragraph, how the second example needs to be used if the Testcontainers `DockerClient` should perform the image pull. And in which cases is this required?

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/JibDockerClient.java:28]: `@Unstable`, to allow us to merge it, but have more flexibility to improve it in the future?

File: docs/features/jib.md
Comment (kiview) [inline docs/features/jib.md:3]:
```suggestion
[Jib](https://github.com/GoogleContainerTools/jib/tree/master/jib-core) is a library for building Docker images. 
You can use it as an alternative to Testcontainers default `DockerfileBuilder`.
```

File: core/src/test/java/org/testcontainers/containers/JibTest.java
Comment (eddumelendez) [reply core/src/test/java/org/testcontainers/containers/JibTest.java:34]: done!

File: docs/features/jib.md
Comment (eddumelendez) [reply docs/features/jib.md:29]: I have added a description for both examples and pointing to Jib's documentation

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/containers/JibDockerClient.java:30]: done!

File: core/src/test/java/org/testcontainers/containers/JibTest.java
Comment (eddumelendez) [reply core/src/test/java/org/testcontainers/containers/JibTest.java]: Given we can set dockerClient in `Jib.from` and `Containerizer.to` as part of `containerize`, constructor `JibImage(Consumer<JibContainerBuilder> jibContainer)` was not added.

Comment (eddumelendez) [reply core/src/test/java/org/testcontainers/containers/JibTest.java]: This is done in the last commit. Description was also updated

File: core/build.gradle
Comment (bsideup) [inline core/build.gradle:78]: To avoid a direct dependency on Jib:
```suggestion
    implementationOnly 'com.google.cloud.tools:jib-core:0.22.0'
```

File: core/src/main/java/org/testcontainers/containers/JibImage.java
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/JibImage.java:54]: nit: `@SneakyThrows` would allow removing these :)

Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/containers/JibImage.java]: my eyes were hurting... thanks for the reminder! :)

File: core/build.gradle
Comment (eddumelendez) [reply core/build.gradle:78]: with `implementation` there is no direct dependency, unlike `api`. User should add the dependency explicitly along with testcontainers in order to make it work as suggested by the docs. Am I missing something?

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/containers/JibDockerClient.java:67]: I know what to do tomorrow :)

File: core/build.gradle
Comment (kiview) [reply core/build.gradle:78]: I think Maven users would still get the transitive dependency, since Maven does not know about the distinction between `api` and `implementation`. At least I used to believe this, is it still correct?

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (kiview) [reply core/src/main/java/org/testcontainers/containers/JibDockerClient.java:67]: Do we still have to tackle this before merging, or can we do it as a future improvement?

File: docs/features/jib.md
Comment (kiview) [inline docs/features/jib.md:11]:
```suggestion
The Testcontainers library JAR will not automatically add a `jib-core` JAR to your project. Minimum version required is `com.google.cloud.tools:jib-core:0.22.0`.
```

File: core/src/main/java/org/testcontainers/containers/BaseConfiguration.java
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/BaseConfiguration.java:7]: Is this interface supposed to be part of this PR? Where is it used?

**Review by kiview — APPROVED** [review]
I think this is good to go after resolving the remaining discussion points 👍 

Through having `JibContainerBuilder` as part of the constructor of `JibImage` we add the Jib types to our public API, but I think this is ok in this case, since the author of `jib-core` is the owner of the Jib project and I don't see a big benefit in decoupling for us here.

File: core/src/main/java/org/testcontainers/containers/BaseConfiguration.java
Comment (eddumelendez) [reply core/src/main/java/org/testcontainers/containers/BaseConfiguration.java:7]: it was leaked. will remote it

File: core/build.gradle
Comment (eddumelendez) [reply core/build.gradle:78]: the pom.xml produced doesn't container any entry for jib.

Comment (kiview) [reply core/build.gradle:78]: Something I found while looking into the best practices around this, could we explore using Gradle's feature variant capabilities for this, and it should provide the best experience for Maven and for Gradle users?
https://docs.gradle.org/current/userguide/feature_variants.html

Comment (kiview) [reply core/build.gradle:78]: Unfortunately, we can't use feature variants with our current Gradle build and publication code (something to look into in the future). We looked at how `implementation` currently behaves for Gradle and Maven consumers and this means we won't have `jib` in the published POM and hence consumers won't get this dependency automatically (double checked this with local build).

Gradle themselves also recommend against using `compileOnly` for modelling optional dependencies. 

We think for now it's ok to go ahead with `implementation` and consider the feature variants approach in the future.

**Review by kiview — DISMISSED** [review]

File: core/src/main/java/org/testcontainers/containers/JibDockerClient.java
Comment (kiview) [reply core/src/main/java/org/testcontainers/containers/JibDockerClient.java:67]: We will mark this with a `TODO` and improve it in a follow up. We don't expect this to be a blocker for early adopters.

**Review by kiview — DISMISSED** [review]

File: core/build.gradle
Comment (bsideup) [reply core/build.gradle:79]: 1. let's not use any Gradle-specific features, or otherwise we will have "fun" time dealing with user reports where the behavior will be dramatically different if they are coming from Maven or Gradle 😅
2. `implementation` is considered a transitive non-API dependency, "implementation detail", while we want to _link_ against it, but it is not needed for TC to work.
3. I just remembered that we have `provided` scope configured ([including publishing](https://github.com/testcontainers/testcontainers-java/blob/65b6aeb40922ea9d0e5e6fa9e607cfbde24512ae/gradle/publishing.gradle#L87)) for this particular use case: https://github.com/testcontainers/testcontainers-java/blob/65b6aeb40922ea9d0e5e6fa9e607cfbde24512ae/modules/selenium/build.gradle#L6

Comment (kiview) [reply core/build.gradle]: 1. I can understand this concern and the effort of staying up-to-date with changes around those features. Dealing with user issues can indeed be challenging if we are not deep in the mental model of Gradle (and we already saw while looking into this topic, that it can be a rabbit hole).
2. True, regarding Gradle semantics, `implementation` is not correct for achieving `provided`
3. Right 😅 Wasn't aware of this anymore (which I think also points at us trying to remove complexity of the Gradle build rather than increasing it). So using `provided` here seems straight forward and consistent within the context of our build. I'd say we should use it here, WDYT @eddumelendez ?

Comment (eddumelendez) [reply core/build.gradle]: @kiview already applied and merged 😄

**Pull request merged**
